### PR TITLE
[TRITON] mHC-pre: Manifold-constrained Hyper Connection 

### PR DESCRIPTION
**Pull request by anhminhnguyenhoang**
Co-authors: @waqahmed-amd-fi @anhminhnguyenhoang

## Motivation

Implements **mHC (manifold-constrained Hyper Connection)** ([arXiv:2512.24880v2](https://arxiv.org/abs/2512.24880v2)) as fused Triton kernels for AMD Instinct GPUs (MI300X / MI355X).

mHC constrains dynamic residual matrices to the Birkhoff polytope via iterative Sinkhorn-Knopp (SK) normalization to prevent gradient explosions in multi-stream residual connections.

## Technical Details

`mhc()` fuses one mHC layer's three streams (`x -> phi_pre, phi_post, phi_res`) into a Triton entry point with two paths chosen at config-load time by the tuned `NUM_KSPLIT`:

1. **`NUM_KSPLIT == 1`** -- single fused kernel `_mhc_fused_kernel` runs RMSNorm, the 3-stream phi projection, per-stream activation (sigmoid+bias for pre/post, in-kernel log-domain Sinkhorn-Knopp for res), and the pre-stream apply step `layer_input = pre_mix * x` inline.
2. **`NUM_KSPLIT > 1`** -- split-K GEMM `_mhc_fused_split_kernel` writes per-K partials into a unified `(NUM_KSPLIT, M, n+n+n^2)` accumulator (one MFMA pass over `x` covers all 3 streams); `_mhc_reduce_apply_kernel` over `(M_blocks, C_blocks)` then reduces partials, runs all 3 stream activations (with sinkhorn), and the apply step in one launch. A `RES_PID_C` constexpr keeps post/res streams pinned to a single workgroup (one CU) when `C` fits one tile.

`sinkhorn_iters` controls the in-kernel Sinkhorn-Knopp iteration count (default 20; `0` returns raw `H^res` logits, skipping SK).

**Layout**:

- Kernels: `aiter/ops/triton/_triton_kernels/fusions/mhc.py` (`_mhc_fused_kernel`, `_mhc_fused_split_kernel`, `_mhc_reduce_apply_kernel`).
- API: `aiter/ops/triton/fusions/mhc.py` — `mhc()`.
- Configs: `aiter/ops/triton/configs/{arch}-MHC_FUSED_SINKHORN-C=*.json` (gfx942/gfx950 buckets: C ∈ {128, 512, 1024, 4096, 32768}; each M-bucket encodes `BLOCK_M/N/K/C`, `NUM_KSPLIT`, `num_warps`, `num_stages`, `waves_per_eu`).
- Reference + tests: `op_tests/triton_tests/utils/mhc_ref.py`, `op_tests/triton_tests/fusions/test_mhc.py`.
- Benchmark: `op_tests/op_benchmarks/triton/bench_mhc.py` (`--with-hip` for side-by-side; needs `--dtype bf16`, n=4, C ≥ 512).

**Further docs:** https://amd.atlassian.net/wiki/spaces/AIG/pages/1388423373/Triton+Kernels+mHC

## Test Plan

`op_tests/triton_tests/fusions/test_mhc.py` (**315 tests**): correctness vs PyTorch reference (bf16/fp16) for both inline (`NUM_KSPLIT==1`) and split-K (`NUM_KSPLIT>1`) branches, in-kernel Sinkhorn-Knopp convergence + doubly-stochastic check, edge cases (zero input, large values, epsilon / alpha / sinkhorn_iters sweeps).

## Test Setup

Replace 'your_NTID' as needed:

```
docker run -it --rm --name mhc-dev --device=/dev/kfd --device=/dev/dri \
  --group-add video --ipc=host --cap-add=SYS_PTRACE --security-opt seccomp=unconfined \
  -v /home/your_NTID/aiter:/workspace/aiter -w /workspace/aiter \
  amdsiloai/pytorch-xdit:v26.4 \
  bash

# inside docker, remove the preinstalled amd_aiter repository and navigate to the one in your workspace
python setup.py develop
pytest op_tests/triton_tests/fusions/test_mhc.py -v -s
python op_tests/op_benchmarks/triton/bench_mhc.py --with-hip -metric all

# from host
docker exec mhc-dev bash -c "cd /workspace/aiter && \
  python op_tests/op_benchmarks/triton/bench_mhc.py --with-hip -metric all"
```

## Test Result

315 / 315 pytest pass. Tolerances: 1e-2 small C, 5e-2 large C (bf16 accumulation).

### gfx950 (MI355X)

Environment: Ubuntu 24.04.4, ROCm 7.12.0, PyTorch 2.9.1 (built against ROCm 7.12.60610), **Triton 3.6.0**, image `amdsiloai/pytorch-xdit:v26.4`.

```
          M    n        C  triton_time(ms) ()  hip_time(ms) ()  triton_throughput(TFLOPS) ()  hip_throughput(TFLOPS) ()  triton_bandwidth(GB/s) ()  hip_bandwidth(GB/s) ()  triton_arithmetic_intensity(FLOP/byte) ()  hip_arithmetic_intensity(FLOP/byte) ()
0    1024.0  4.0    512.0            0.020183         0.016065                      5.757620                   7.305041                 474.268559              598.094582                                  12.167543                               12.167543
1    2048.0  4.0    512.0            0.020808         0.020256                     11.290713                  11.476962                 923.167159              938.983936                                  12.230377                               12.230377
2    4096.0  4.0    512.0            0.024919         0.031051                     17.324461                  15.019292                1385.066740             1226.109236                                  12.262038                               12.262038
3    8192.0  4.0    512.0            0.025388         0.043662                     36.694217                  21.328914                2999.368731             1737.379589                                  12.277930                               12.277930
4   16384.0  4.0    512.0            0.037381         0.071595                     49.825171                  26.014400                4063.776819             2118.781610                                  12.285892                               12.285892
5    1024.0  4.0   4096.0            0.027825         0.038610                     32.628283                  23.570331                2735.764668             1978.944752                                  11.912832                               11.912832
6    2048.0  4.0   4096.0            0.038498         0.054349                     47.314866                  33.461711                3952.588969             2792.918553                                  11.974531                               11.974531
7    4096.0  4.0   4096.0            0.068677         0.097873                     52.899779                  37.178745                4408.121771             3093.013013                                  12.005621                               12.005621
8    8192.0  4.0   4096.0            0.148375         0.194224                     48.899093                  37.427322                4069.535670             3114.827131                                  12.021226                               12.021226
9   16384.0  4.0   4096.0            0.281813         0.340491                     51.274346                  42.864280                4256.896833             3545.912907                                  12.029044                               12.029044
10   1024.0  4.0  32768.0            0.176901         0.187038                     40.862182                  38.760063                3439.896961             3262.154544                                  11.880859                               11.880859
11   2048.0  4.0  32768.0            0.308764         0.355717                     47.156672                  40.853869                3962.918740             3420.858016                                  11.942414                               11.942414
12   4096.0  4.0  32768.0            0.518628         0.714735                     55.971707                  40.635560                4681.724568             3393.896267                                  11.973432                               11.973432
13   8192.0  4.0  32768.0            0.954769         1.355924                     60.754902                  42.828688                5073.076328             3565.385508                                  11.989001                               11.989001
14  16384.0  4.0  32768.0            1.805885         2.587059                     64.224246                  44.871964                5351.196291             3730.574357                                  11.996801                               11.996801
```

### gfx942 (MI300X)

Environment: Ubuntu 22.04, ROCm 7.2.2, PyTorch 2.9.1+rocm7.2.2 (`amdsiloai/pytorch-xdit:v26.4`), **Triton 3.6.0**.

```
          M    n        C  triton_time(ms)  hip_time(ms)  triton_throughput(TFLOPS)  hip_throughput(TFLOPS)  triton_bandwidth(GB/s)  hip_bandwidth(GB/s)  triton_arithmetic_intensity(FLOP/byte)  hip_arithmetic_intensity(FLOP/byte)
0    1024.0  4.0    512.0         0.039681      0.027936                   2.936494                4.171070              241.338273           342.802978                               12.167543                            12.167543
1    2048.0  4.0    512.0         0.038639      0.029071                   6.031368                8.016443              493.146510           655.453476                               12.230377                            12.230377
2    4096.0  4.0    512.0         0.046535      0.030047                  10.015946               15.512099              816.825529          1265.050621                               12.262038                            12.262038
3    8192.0  4.0    512.0         0.050351      0.045053                  18.513715               20.690832             1507.885683          1685.205247                               12.277930                            12.277930
4   16384.0  4.0    512.0         0.059619      0.073868                  31.271375               25.239185             2545.307771          2054.322630                               12.285892                            12.285892
5    1024.0  4.0   4096.0         0.068254      0.045774                  13.321512               19.863819             1118.248894          1667.430419                               11.912832                            11.912832
6    2048.0  4.0   4096.0         0.077456      0.066726                  23.477754               27.253139             1960.640777          2275.925306                               11.974531                            11.974531
7    4096.0  4.0   4096.0         0.119564      0.125503                  30.418737               28.979274             2533.707939          2413.808881                               12.005621                            12.005621
8    8192.0  4.0   4096.0         0.246793      0.292962                  29.473979               24.829062             2451.827985          2065.435053                               12.021226                            12.021226
9   16384.0  4.0   4096.0         0.510828      0.543159                  28.479143               26.783950             2367.531615          2226.606647                               12.029044                            12.029044
10   1024.0  4.0  32768.0         0.271851      0.287498                  26.672825               25.221164             2245.024988          2122.840117                               11.880859                            11.880859
11   2048.0  4.0  32768.0         0.535561      0.533420                  27.078275               27.186960             2267.403758          2276.504488                               11.942414                            11.942414
12   4096.0  4.0  32768.0         1.060194      1.071146                  27.357386               27.077669             2284.840790          2261.479290                               11.973432                            11.973432
13   8192.0  4.0  32768.0         2.027440      2.053475                  28.611585               28.248833             2386.486130          2356.229046                               11.989001                            11.989001
14  16384.0  4.0  32768.0         4.308759      4.068728                  26.925745               28.514205             2244.410358          2376.817356                               11.996801                            11.996801
```

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: aiter/ops/triton/fusions/mhc.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+    bias: torch.Tensor,
+    n: int,
+    eps: float = 1e-6,
+    hc_pre_eps: float = 0.0,
+    hc_post_mult_value: float = 2.0,
+    sinkhorn_iters: int = 20,
+    config: Optional[dict] = None,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    """
+    Single entry point ``mhc()`` runs equations 14-19 plus the layer_input apply
```
Comment (vgokhale) [inline aiter/ops/triton/fusions/mhc.py]: Can you change this summary a bit? May be an ascii block diagram instead? This seems to refer to equations in the paper.

**Comment by vgokhale** [comment]
I have a question on the choice of fusions. Here is what mhc pre does

1) mix logits gemm and divide by RMS
2) slice [0:4] -> sigmoid -> weighted sum with residual -> pre mix layer output
3) slice [4:8] -> sigmoid -> post mix
4) slice [8:24] -> softmax -> sinkhorn

This PR is doing splitk for 1. But IIUC, it is then parallelizing along 2/3/4. But that means it redundantly reads the split partial products.

Why not parallelize along the token and hidden dimensions instead? So

Kernel 1: Produce mix logits splitk

Kernel 2: Each pid reads its slice of splitk partial products to produce final mix logits output. Then does all 3 steps 2-4 above. With step 2 being done only for a BLOCK_H of output hidden size columns.

**Comment by vgokhale** [comment]
LGTM. One last request - not necessary right now but can you try to shrink the config files to may be C = large and small? This is mainly for maintainability.

**Review by vgokhale — APPROVED** [review]

**Comment by anhminhnguyenhoang** [comment]
@vgokhale surely I'll revisit this and make the update on the mhc post PR
